src/core/infer.ml

```diff
@@ -1,216 +1,201 @@
-         (no?) chance that it can be a kind. *)
-      end
-
-  (*  c ⊢ a ⇐ Type    c, x:a ⊢ t ⇒ b   b ≠ Kind
-     --------------------------------------------
-             c ⊢ Abst(a,t) ⇒ Prod(a,b)          *)
-  | Abst(a,t) ->
-      check p c a mk_Type;
-      let x,t,c' = Ctxt.unbind c a None t in
-      let b = infer p c' t in
-      begin match unfold b with
-      | Kind ->
-          wrn None "Abstraction on [%a] is not allowed." Print.pp_term t;
-          raise NotTypable
-      | _ -> mk_Prod(a, Bindlib.unbox (Bindlib.bind_var x (lift b)))
-      end
-
-  (*  c ⊢ t ⇒ Prod(a,b)    c ⊢ u ⇐ a
-     ------------------------------------
-         c ⊢ Appl(t,u) ⇒ subst b u      *)
-  | Appl(t,u) ->
-      (* [get_prod f typ] returns the domain and codomain of [t] if [t] is a
-         product. If [t] is a metavariable, then it instantiates it with a
-         product and calls [get_prod f typ] again. Otherwise, it calls [f
-         typ]. *)
-      let rec get_prod f typ =
-        if !log_enabled then log_infr "get_prod %a" pp_term typ;
-        match unfold typ with
-        | Prod(a,b) -> (a,b)
-        | Meta(m,_) -> set_to_prod p m; get_prod f typ
-        | _ -> f typ
+      let a = try Ctxt.type_of x ctx with Not_found -> assert false in
+      (t, a)
+  | Symb s -> (t, !(s.sym_type))
+  | Plac (true, name) ->
+      let m = LibMeta.make pb ?name ctx mk_Type in
+      (m, mk_Type)
+  | Plac (false, name) ->
+      let mt = LibMeta.make pb ctx mk_Type in
+      let m = LibMeta.make pb ?name ctx mt in
+      (m, mt)
+  (* All metavariables inserted are typed. *)
+  | (Meta (m, ts)) as t ->
+      let rec ref_esubst i range =
+        (* Refine terms of the explicit substitution. *)
+        if i >= Array.length ts then range else
+          match unfold range with
+          | Prod(ai, b) ->
+              ts.(i) <- force pb ctx ts.(i) ai;
+              let b = Bindlib.subst b ts.(i) in
+              ref_esubst (i + 1) b
+          | _ ->
+              (* Meta type must be a product of arity greater or equal
+                 to the environment *)
+              assert false
+      in
+      let range = ref_esubst 0 !(m.meta_type) in
+      (t, range)
+  | LLet (t_ty, t, u) ->
+      (* Check that [a] is a type, and refine it. *)
+      let t_ty, _ = type_enforce pb ctx t_ty in
+      (* Check that [t] is of type [a'], and refine it *)
+      let t = force pb ctx t t_ty in
+      (* Unbind [u] and get new context with [x: t_ty ≔ t] *)
+      let x, u, ctx = Ctxt.unbind ctx t_ty (Some t) u in
+      (* Infer type of [u'] and refine it. *)
+      let u, u_ty = infer pb ctx u in
+      let u_ty = Bindlib.(u_ty |> lift |> bind_var x |> unbox) in
+      let u_ty = Bindlib.subst u_ty t in
+      let u = Bindlib.(u |> lift |> bind_var x |> unbox) in
+      (mk_LLet (t_ty, t, u), u_ty)
+  | Abst (dom, b) ->
+      (* Domain must by of type Type, we don’t use [type_enforce] *)
```

Comment:
Typo: by -> be

Reply:
Why don't you use type_enforce here? You do it in the let case.

Reply:
`type_enforce` would allow `dom` to be of type `TYPE` or `KIND` so it would allow an expression `\s:TYPE, t`, which shouldn't be allowed because we don't have a product `(Kind, _, _)` in lambdapi, right?

Reply:
I see. Perhaps just remove ", we don’t use [type_enforce]" to avoid any confusion.